doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
+ * `size`: an integer, the (apparent) size of the file (`st_size`)
+ * `modified`: a float, the mtime of the file (`st_mtim`)
+ * `tag`: a value type: the current 'transaction tag' of the file, with the
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
+ * `entries`: for directories: the contents of the directory (see below)
+ * `content`: for regular text files, the content (if less than 4096 bytes)
+ * `target`: for symbolic links: the target of the link (a string)
+
+The `entries` attribute is only reported on directories, and only if the
+contents of the directory can be read.  It is an object where each key is the
+name of a file in the directory, and each value is an object describing that
+file, using the same attributes as above.  `entries` is not reported on entries
+(ie: no recursive listing).
+
+The attributes to be reported can be controlled by setting the `attrs` option.
+
+If the provided `path` is a symbolic link, it is followed, unless
+`follow: false` is specified.  If symbolic links are present in the directory,
+they are always reported directly (ie: not followed).  `follow: false` is not
+supported in combination with `watch: true` but you probably don't need it:
+watch mode doesn't only watch the final path component, but all components
+leading up to it, including any symlink that have been followed.  That means
+that you can watch `/etc/localtime` and find out about if the link changed, or
+if the contents of the file itself changed, or was deleted or replaced, or also
+if `/usr`, `/etc`, or any other directories or intermediate links were changed,
+deleted, or moved out of the way.
+
+(TODO: providing `watch: true, follow: false` would be possible with an
+addition to the `systemd_ctypes` `PathWatch` API).
+
+The life-cycle of the channel works like this:
+
+ - When opening the channel, if a file at the named `path` cannot be opened and
+   `close-on-error` is true, opening fails (`close` with `problem`).
+
+ - Otherwise, `ready` is sent and the value representing the current state is
+   reported, as above.  Only when the current value reaches a state where
+   `partial` is unset can the initial value be read.
+
+ - If `watch` mode is false then `done` is sent immediately, and the channel is
+   closed.
+
+ - Otherwise (`watch: true`), the channel remains open.  When any changes are
+   detected, they are sent as updates to the value of the channel.  If the path
+   itself cannot be opened at any point, and `close-on-error` is set, the channel
+   will close (with `problem`).  Otherwise, the value may become `null`, but
+   the channel will never close itself.
```

Comment:
Will there ever be `partial` updates in watch mode? I guess/hope not, but this could be clarified here.

Reply:
I'm not sure, and I intend to leave the question unanswered for now.

This is all going to be more meaningful if/when we try to split updates into smaller chunks to enable flow control.

Reply:
There is one spot where `partial` will currently get used for watching, in fact:

If you transition from an error state to a directory then you'll send a patch like:

```js
{'problem': null, 'message': null, 'entries': { ... }}
```

and if you transition from directory to error:

```
{'problem': 'bzzt', 'message': 'uh oh', 'entries': null }}
```

but if you transition from one directory to another directory in the same location (as may happen if a symlink gets changed, for example) then you need to transition directly from reporting the attributes of one directory to reporting the attributes of a second directory.

That presents a problem.  Due to the way merge patching works, unless you want to explicitly send `"entries": {"foo": null, "bar": null}` updates for a potentially very large number of files, you can't replace the "entries" dict with a new one in one step.

For this case we send a message that sets `{"entries": null, "partial": true}` first, and then a second message which sets the desired end value of `"entries"`.

That's the existing implementation and it's a place where the client library should fail to report the intermediate data to its user.